iothub/device/src/Transport/AmqpClientConnectionPool.cs

```diff
@@ -0,0 +1,155 @@
+        {
+            if (Logging.IsEnabled) Logging.Enter(this, $"{nameof(AmqpClientConnectionPool)}.{nameof(GetClientConnection)}");
+
+            if (deviceClientEndpointIdentity is DeviceClientEndpointIdentitySasSingle)
+            {
+                return GetIoTHubSasSingleClientConnection(deviceClientEndpointIdentity);
+            }
+            else if (deviceClientEndpointIdentity is DeviceClientEndpointIdentityX509)
+            {
+                return GetIoTHubX509ClientConnection(deviceClientEndpointIdentity);
+            }
+            else if (deviceClientEndpointIdentity is DeviceClientEndpointIdentitySASMux)
+            {
+                return GetIoTHubSasMuxClientConnection(deviceClientEndpointIdentity);
+            }
+            else if (deviceClientEndpointIdentity is DeviceClientEndpointIdentityIoTHubSas)
+            {
+                return GetIoTHubSasClientConnection(deviceClientEndpointIdentity);
+            }
+            else
+            {
+                throw new NotImplementedException();
+            }
+        }
+        internal void RemoveClientConnection(DeviceClientEndpointIdentity deviceClientEndpointIdentity)
+        {
+            if (Logging.IsEnabled) Logging.Enter(this, $"{nameof(AmqpClientConnectionPool)}.{nameof(RemoveClientConnection)}");
+
+            if (deviceClientEndpointIdentity is DeviceClientEndpointIdentitySasSingle)
+            {
+                if (singleSasPool.ContainsKey(deviceClientEndpointIdentity))
+                {
+                    AmqpClientConnection amqpClientConnection;
+                    singleSasPool.TryRemove(deviceClientEndpointIdentity, out amqpClientConnection);
+
+                    if (amqpClientConnection != null)
+                    {
+                        if (Logging.IsEnabled) Logging.Info(this, $"{nameof(AmqpClientConnectionPool)}.{"Removed from singleSasPool:"}.{nameof(amqpClientConnection)}");
+                    }
+                }
+            }
+            else if (deviceClientEndpointIdentity is DeviceClientEndpointIdentityX509)
+            {
+                if (x509Pool.ContainsKey(deviceClientEndpointIdentity))
+                {
+                    AmqpClientConnection amqpClientConnection;
+                    x509Pool.TryRemove(deviceClientEndpointIdentity, out amqpClientConnection);
+
+                    if (amqpClientConnection != null)
+                    {
+                        if (Logging.IsEnabled) Logging.Info(this, $"{nameof(AmqpClientConnectionPool)}.{"Removed from X509Pool:"}.{nameof(amqpClientConnection)}");
+                    }
+                }
+            }
+            if (Logging.IsEnabled) Logging.Exit(this, $"{nameof(AmqpClientConnectionPool)}.{nameof(RemoveClientConnection)}");
+        }
+
+
+        private AmqpClientConnection GetIoTHubSasSingleClientConnection(DeviceClientEndpointIdentity deviceClientEndpointIdentity)
```

Comment:
Suggest to rename function to `CreateIoTHubSasSingleClientConnection`.

Reply:
This is a private function and it is supposed the GET the connection from the underlying pool (if any). I prefer GET.

---
In reply to: [251626740](https://github.com/Azure/azure-iot-sdk-csharp/pull/770#discussion_r251626740) [](ancestors = 251626740)

Reply:
This function is implemented as "create a connection if it doesn't exist otherwise return null". Normally GET means "return  a connection if it exists", and shouldn't change current status.

Reply:
Agreed, I will fix it in the MuxPool (this pool is already obsolete)

---
In reply to: [251667364](https://github.com/Azure/azure-iot-sdk-csharp/pull/770#discussion_r251667364) [](ancestors = 251667364)